common/src/progenitor_operation_retry.rs

```diff
@@ -0,0 +1,158 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use futures::Future;
+use slog::warn;
+use slog::Logger;
+
+use crate::api::external::Error;
+use crate::backoff::retry_notify;
+use crate::backoff::retry_policy_internal_service;
+use crate::backoff::BackoffError;
+
+#[derive(Debug)]
+pub enum ProgenitorOperationRetryError<E> {
+    /// Nexus determined that the operation will never return a known result
+    /// because the remote server is gone.
+    Gone,
+
+    /// Attempting to check if the retry loop should be stopped failed
+    GoneCheckError(Error),
+
+    /// The retry loop progenitor operation saw a permanent client error
+    ProgenitorError(progenitor_client::Error<E>),
+}
+
+/// Retry a progenitor client operation until a known result is returned, or
+/// until something tells us that we should stop trying.
+///
+/// Saga execution relies on the outcome of an external call being known: since
+/// they are idempotent, reissue the external call until a known result comes
+/// back. Retry if a communication error is seen, or if another retryable error
+/// is seen.
+///
+/// During the retry loop, call the supplied `gone_check` function to see if the
+/// retry loop should be aborted: in the cases where Nexus can _know_ that a
+/// request will never complete, the retry loop must be aborted. Otherwise,
+/// Nexus will indefinitely retry until some known result is returned.
+///
+/// Note that retrying is only valid if the `operation` itself is idempotent.
+pub struct ProgenitorOperationRetry<
+    T,
+    E: std::fmt::Debug,
+    F: FnMut() -> Fut,
+    Fut: Future<Output = Result<T, progenitor_client::Error<E>>>,
+    BF: FnMut() -> BFut,
+    BFut: Future<Output = Result<bool, Error>>,
+> {
+    operation: F,
+
+    /// If Nexus knows that the supplied operation will never successfully
+    /// complete, then `gone_check` should return true.
+    gone_check: BF,
+}
+
+impl<T, E, F, Fut, BF, BFut> ProgenitorOperationRetry<T, E, F, Fut, BF, BFut>
+where
+    E: std::fmt::Debug,
+    F: FnMut() -> Fut,
+    Fut: Future<Output = Result<T, progenitor_client::Error<E>>>,
+    BF: FnMut() -> BFut,
+    BFut: Future<Output = Result<bool, Error>>,
+{
+    pub fn new(operation: F, gone_check: BF) -> Self {
+        Self { operation, gone_check }
+    }
+
+    pub async fn run(
+        mut self,
+        log: &Logger,
+    ) -> Result<T, ProgenitorOperationRetryError<E>> {
+        retry_notify(
```

Comment:
It looks like when you call `run` inside the `create_region` future inside `ensure_region_in_dataset`, there is a [`retry_notify`](https://github.com/oxidecomputer/omicron/pull/5789/files#diff-93ab1121effeed1199aef2b1d99ea8a00da10b6998c21aba461d15269a1f9c44R179-R182) call around the `create_region` future. This creates two backoff loops rather than one, and seems like a really easy mistake to make. Instead, what if we didn't actually call retry instead of `run` here but relied on the outer retry loop. I think that would simplify this a lot as well, as now you wouldn't have to worry about the progenitor errors. The outer loop would call a future that only does the gone check once and returns an error, rather than looping.

Reply:
In this case, the double backoff looping was intentional: the outer retry (in nexus/src/app/crucible.rs) is to poll until a desired state change, and the inner retry (in common/src/progenitor_operation_retry.rs) is to paper over transient network problems. I think I still like this separation, though I don't grok the simplification you're referring to yet. Going to hack around with this.

Reply:
If this makes sense, and your hacking didn't yield anything, I'm fine leaving it as is for now.